Feature/react part2 (#1333)
Techtonica/curriculum: react-js/react-part-2.md

# React Part 2

## Projected Time

3 hours 30 mins

- Lesson: 1 hour
- Guided Practice: 1 hour
- Independent Practice: 1 hour 30 mins

### Prerequisites

- [React](react.md)

### Motivation

> In introduction to [React](/react.md) we discussed about how React makes websites interactive and reduces jQuery code. Now the problem arises if we want to store some information which intent to change and due to change of that information the UI changes i.e the page is re-rendered. Thus, component state is used for such cases.

Reviewer: ```suggestion
In the introduction to [React](/react.md), we discussed how React makes websites interactive and removes the need for jQuery code. 

Now the problem arises: How do we store and change component information we want in the UI?  For example, maybe you want a button's background color to change once it's been clicked.  The solution is to use React `state`.
```

Using jQuery and React together is a big no-no, so let's not give them that idea.

### Objectives

- Component state
- State vs Props
- State is Immutable

### Lesson

- [Introduction to component state in React (video)](https://www.youtube.com/watch?v=e5n9j9n83OM)
- [Component state in React slides](https://docs.google.com/presentation/d/1Trj_7vWYe79_mDboDxN9rdT94yjY5Dt55xtU1k3dUsw/edit?usp=sharing)

### Common Mistakes & Misconceptions

> 1. The scope of the state is local to the component.

Reviewer: Can you give an example of what will and won't work? For both of these please.  I don't actually know what 'Don't forget to update state immutably.' means.

2. Don't forget to update state immutably.

> 

Reviewer: Can you please include the common mistake of trying to re-render by changing props?  An example of it failing with props and succeeding with state would be important - could you add one to Guided practice?

### Guided Practice

- Define state.
  ```javascript
  state = {};
  ```
- Initialize state inside class constructor.

  ```javascript
  constructor(){
    this.state = {
    message: "Hello world"
    }
  }
  ```

> - State should not be mutated directly i.e. a copy of state should be created and then updated.

Reviewer: ```suggestion
- State should not be mutated directly. A copy of state should be created and then updated using `this.setState`.
```

```javascript
{
  this.state.value < 10 && 
	<button onClick={this.addValue}>Increment</button>
}
{
  this.state.value > 0 &&
    <button onClick={this.subtractValue}>Decrement</button>
}
```

That's all. We have completed our first application in React using React state. To see the complete code [Codepen](https://codepen.io/ashish24_nagpal/pen/jObzXzM).

### Independent Practice

For independent practice, let us create a simple game. Say you are a magician having a magic box and 3 random buttons. Out of 3 only 1 button is correct which reveals the text inside the magic box. Children will come to you and choose a button. If the button is correct, any random text should display which is present inside the box and status message should be displayed as _you win_. If the button chosen is incorrect then only message should be displayed as _you lose_.  
UI should contain -

- A div which is a magic box.
- 3 buttons out of which only 1 is correct.
- A hidden text which will only be displayed when correct button is clicked.
- A message tag which display the status of the game whether the person has _won_ or _lost_ the game.
- A reset button which will reset the state of the game.  

>   **_Hint :-_**

Reviewer: what is the hint?  lines 168 and 169?  Fix the formatting so they appear as sub-bullets.

# React Part 2

## Projected Time

1.5 hours
- Lesson: 30min
- Guided Practice: 30min
- Independent Practice: 30min

### Prerequisites

- [React](react.md)
- [JSX](react.md)
- [Props](react.md)

### Motivation

In introduction to [React](/react.md) we discussed about how React makes websites interactive and reduces jQuery code. Now the problem arises if we want to store some information which intent to change and due to change of that information the UI changes i.e the page is re-rendered. Thus, component state is used for such cases.

### Objectives

- Component state
- State vs Props
- State are Immutable

### Lesson

> [Slides](https://docs.google.com/presentation/d/1Trj_7vWYe79_mDboDxN9rdT94yjY5Dt55xtU1k3dUsw/edit?usp=sharing)

Reviewer 2: Isn't the presentation public ? It's showing me access denied.
@ashishnagpal2498 And please specify what the slide is about.

Author: Yes , it wasn't. I just corrected it, you can view it now. Also I have changed the owner.

Reviewer 2: I need edit access to review the presentation. Until you fix other issues.

Author: Kindly provide me your email @Kundan28

Reviewer 2: I already requested, you might have received an email.

Author: No , I haven't received any mail. May be because I have transferred the ownership. Provide your email so that I can give you the access.

[What is state](https://www.youtube.com/watch?v=e5n9j9n83OM)

> 

Reviewer 2: ```suggestion
```

Reviewer: Please commit Kundan's suggestion?

### Common Mistakes & Misconceptions

1. The scope of the state is local to the component.
2. Don't forget to update state immutably.

> 

Reviewer 2: ```suggestion
```

### Guided Practice

1. Define state.
  - >	``` state = {} ```
2. Initialize state
  - State in class component can be initialized in two ways
    - Directly inside the 

3. State should not be mutated directly i.e. a copy of state should be created and then updated. (already covered in lesson)
    - Incorrect way -
      - > ```this.state.list.push(value)```
    - Correct way -

>       - > ``` this.setState( { ...this.state , list: [ ...this.state.list, value ] } )```  
> #### Now , let us create a simple todo list.

Reviewer 2: ```suggestion
      - > ``` this.setState( { ...this.state , list: [ ...this.state.list, value ] } )```  
      
#### Now , let us create a simple todo list.
```

1. Create a new React application using the command `npx create-react-app .`

> 2. This will create a simple React app , and now we will start making changes to this app.

Reviewer 2: ```suggestion
2. This will create a simple React app, and now we will start making changes to this app.
```

> 1. Create a new React application using the command `npx create-react-app .`
> 2. This will create a simple React app , and now we will start making changes to this app.

Reviewer 2: ```suggestion
1. Create a new React application using the command `npx create-react-app .`

2. This will create a simple React app , and now we will start making changes to this app.
```

> 3. First of all get rid of all the unnecessary stuff showing on DOM , remove all the lines in render of `App.js`. 

Reviewer 2: ```suggestion
3. First of all get rid of all the unnecessary stuff showing on DOM, remove all the lines in render of `App.js`. 
```

>   - ```

Reviewer 2: Remove the unwanted dash(-)

    import React from 'react';
    import './App.css';

    class App extends React.Component {
      render() {	
        return (
          <div className="App">  
          </div>
        );
      }
    }

    export default App;
    ```	
4. Now we will create an input box which will take our task values and a button which add new values to the list.
-  ```
    <input />
    <button type="button" >Add</button>
    ```

> 5. Now we will create state object of the component and assign properties to it.
> #### Note :- All the input box defined in React are controlled by state. In other words the value of input box should be assigned to a property of state object.

Reviewer 2: ```suggestion
5. Now we will create state object of the component and assign properties to it.

#### Note :- All the input box defined in React are controlled by state. In other words the value of input box should be assigned to a property of state object.
```

# React Part 2

## Projected Time

1.5 hours
- Lesson: 30min

> - Guided Practice: 30min

Reviewer 3: Please increase these values as well. There's no way a new person would get through this in 1.5 hours. Have someone try it out to get a time estimate or make your best guess, erring on the side of it being an overestimate.

- Independent Practice: 30min

### Prerequisites

- [React](react.md)
- [JSX](react.md)

> - [Props](react.md)

Reviewer 3: These are prerequisite lessons, not concepts so please turn into a single prereq

### Motivation

In introduction to [React](/react.md) we discussed about how React makes websites interactive and reduces jQuery code. Now the problem arises if we want to store some information which intent to change and due to change of that information the UI changes i.e the page is re-rendered. Thus, component state is used for such cases.

### Objectives

- Component state
- State vs Props
- State are Immutable

### Lesson

- [Introduction to component state in React (video)](https://www.youtube.com/watch?v=e5n9j9n83OM)
- [Component state in React slides](https://docs.google.com/presentation/d/1Trj_7vWYe79_mDboDxN9rdT94yjY5Dt55xtU1k3dUsw/edit?usp=sharing)

### Common Mistakes & Misconceptions

1. The scope of the state is local to the component.
2. Don't forget to update state immutably.

### Guided Practice

1. Define state.
  - >	``` state = {} ```
2. Initialize state
  - State in class component can be initialized in two ways
    - Directly inside the 

> 

Reviewer 3: This is unfinished. Also, I don't understand why you have No. 1 and No. 2 

No. 2 should cover it. Add a code snippet showing this.state initialization in a constructor (with at least one key and value)

Reviewer 3: @Kundan28 can you review more carefully so you don't approve this type of content.

Author: Okay, I'll do the changes.

3. State should not be mutated directly i.e. a copy of state should be created and then updated. (already covered in lesson)
    - Incorrect way -

>       - > ```this.state.list.push(value)```

Reviewer 3: Please use JavaScript syntax for all snippets and make them multi-line

Author: I don't know what exactly JavaScript syntax is. Can you please guide me to it.

Reviewer 3: I googled `github markdown javascript syntax`

and found this doc page: https://guides.github.com/features/mastering-markdown/ and the heading "Syntax Highlighting"

Reviewer 3: This is still missing and the formatting of the code snippets is inconsistent. Please use a formatter like [prettier](https://prettier.io/) to help you if you're having trouble doing it manually.

    - Correct way -

>       - > ``` this.setState( { ...this.state , list: [ ...this.state.list, value ] } )```

Reviewer 3: Please expand on what `setState` does and how it related to re-rendering (this was in the original GitHub description) and is important.

> #### Now , let us create a simple todo list.

Reviewer 3: This is a really big jump. Please guide them through the digital clock example or something easy first like a single true/false value. But this is a great example later for more practice.

Author: Okay @vegetabill , I will add a simpler example. So should I put this content in challenge section ?

Reviewer 3: I would say just omit it. We will have other lessons to cover it.

> 1. Create a new React application using the command `npx create-react-app .`

Reviewer 3: Remove this. We don't want the code to be so complex we need CRA

# React Part 2

## Projected Time

2 hours 30 mins

- Lesson: 1 hour
- Guided Practice: 30 mins
- Independent Practice: 1 hour

### Prerequisites

- [React](react.md)

### Motivation

In introduction to [React](/react.md) we discussed about how React makes websites interactive and reduces jQuery code. Now the problem arises if we want to store some information which intent to change and due to change of that information the UI changes i.e the page is re-rendered. Thus, component state is used for such cases.

### Objectives

- Component state
- State vs Props

> - State are Immutable

Reviewer 3: Typo: State *is* immutable

### Lesson

- [Introduction to component state in React (video)](https://www.youtube.com/watch?v=e5n9j9n83OM)
- [Component state in React slides](https://docs.google.com/presentation/d/1Trj_7vWYe79_mDboDxN9rdT94yjY5Dt55xtU1k3dUsw/edit?usp=sharing)

### Common Mistakes & Misconceptions

1. The scope of the state is local to the component.
2. Don't forget to update state immutably.

### Guided Practice

- Define state.

>   ```

Reviewer 3: All code snippets need to be marked as JavaScript so they will have proper highlighting.

This was an earlier unresolved comment. This is covered under "Syntax Highlighting" on https://guides.github.com/features/mastering-markdown/

  state = {}
  ```
- Initialize state

  - State in class component can be initialized in two ways

    - Directly inside the class component, outside all methods.

      ```
      state = {
        message: "Hello world"
      }
      ```

    - Inside class constructor
      ```
      constructor(){
        this.state = {
          message: "Hello world"
          }
      }

>       ```

Reviewer 3: As this is for beginners, let's just show the second way to do it so please remove the first way on line 48

> - State should not be mutated directly i.e. a copy of state should be created and then updated.

Reviewer 3: Can you change this code to follow the `message` example you start with?

e.g. you don't do `state.message = "new message"`

Reviewer: Why did you resolve this?  Did you fix it?

  - Incorrect way -
    ```
    this.state.list.push(value)
    ```
  - Correct way -
    ```
        this.setState({

>           ...this.state,

Reviewer 3: `setState` will automatically merge so this is redundant to include. Let's remove to keep it simpler and just include the `list` key.

          list: [ ...this.state.list, value ]
        })
    ```

> - You must have understand till now that we should not update state directly, but what special thing _setState_ function does to the state, that the component re-renders automatically. Well, in introduction to React, we have learnt that React follows a special path called the React cycle. So, when state is updated using setState, React goes into the **React cycle** and then the changes are compared from the previous **React virtual DOM element**, changes are done accordingly and those changes are reflected in the UI.

Reviewer 3: This first sentence has confusing grammar. Please simplify to:

> The important thing to remember about `setState` is that is causes the component to re-render

> #### Now, let us create a simple todo list.

Reviewer 3: This example is way too complex for this lesson. Please change the guided practice to be a walkthrough of how to implement the counter you mention in the independent pratice

> 1. First of all get rid of all the unnecessary stuff showing on DOM, remove all the lines in render of `App.js`.

Reviewer 3: This looks like leftover from using CRA. Please remove

   ```

>     import React from 'react';

Reviewer 3: Can you create a starting CodePen for this with the right libraries and syntax? For example, there will be no `import` or `export`.

here is a tiny example: https://codepen.io/vegetabill/pen/rNOMZGK

    import './App.css';

    class App extends React.Component {
      render() {
        return (
          <div className="App">
          </div>
        );
      }
    }

    export default App;
   ```

2. Now we will create an input box which will take our task values and a button which add new values to the list.

   ```
    <input />
    <button type="button" >Add</button>
   ```

3. Now we will create state object of the component and assign properties to it.

> #### Note :- All the input box defined in React are controlled by state. In other words the value of input box should be assigned to a property of state object.

Reviewer 3: This is not a heading. If you want to bold a note, use bold.

   ```
   <ol>
     {this.state.items.map((item,index) => (
       <li key={index}>
        {item}
       </li>
      ))}
    </ol>
   ```

8. Finally your `App.js` will look like -

    ```
    import React from 'react';
    import './App.css';

    class App extends React.Component {
     state = {
       items: [],
       value:""
     }

    // Function to handle input value of the text field
     onChangeHandler = (event) => {
       this.setState({
         value: [event.target.value]
       })
     }

    // Function to add new values to items array
     addItem = () => {
       this.setState({
         value: "",
         items: [...this.state.items,this.state.value]
       })
     }

     render() {
       return (
         <div className="App">
           <input value={this.state.value} onChange={this.onChangeHandler} />
           <button type="button" onClick={this.addItem} >Add</button>
           <ol>
             {this.state.items.map((item,index) => (
               <li key={index}>
                 {item}
               </li>
              ))}
            </ol>
          </div>
        );
      }
    }

    export default App;
    ```

### Independent Practice

> - Create an increment/decrement counter which includes click functionality. On click of increment button adds 1 to the value and decrement button will subtract the number by 1. Initially the number will be 0.

Reviewer 3: Since this is now the guided practice, please come up with another one for independent practice. A list of items would be good, as shown above.

## Projected Time

> 2.5 hours

Reviewer 2: ```suggestion
2 hours 30 minutes
```

- Lesson: 1 hour
- Guided Practice: 30 mins
- Independent Practice: 1 hour

### Prerequisites

> - [React](react.md) - To start with this section you must be thorough about *what is React*, *what are components in React*, *JSX* and *props*. 

Reviewer 3: Please just list the lesson link, not the specific sections related

### Motivation

In the introduction to [React](/react.md), we discussed how React makes websites interactive and removes the need for jQuery code.  

> Now the problem arises: How do we store and change component information we want in the UI?  For example, maybe you want a button's background color to change once it's been clicked.  The solution is to use React `state`.

Reviewer 2: ```suggestion
Now the problem arises: How do we store and change component information we want in the UI?  For example, maybe you want a button's background color to change once it's been clicked. The solution is to use React `state`.
```

### Guided Practice

- Define state.
  ``` 
  state = {} 
  ```
- Initialize state
  - State in class component can be initialized in two ways
    - Directly inside the class component, outside all methods.
      ``` 
      state = {
        message: "Hello world"
      }
      ```

    - Inside class constructor
      ```
      constructor(){
        this.state = {
          message: "Hello world"
          }
      }
      ```

> - State should not be mutated directly i.e. a copy of state should be created and then updated. (already covered in lesson)

Reviewer 3: I think the `(already covered...` is leftover from an older version. It has not already been covered so it's confusing. Please remove this.

    - Correct way -
      ``` 
	  this.setState({
		  ...this.state,
		  list: [ ...this.state.list, value ]
	  })
	  ```
- You must have understand till now that we should not update state directly, but what special thing *setState* function does to the state, that the component re-renders automatically. Well, in introduction to React, we have learnt that React follow a special path called the React cycle. So, when state is updated using setState, React goes into **shouldComponentUpdate** and **componentWillUpdate** method (in case of state update and an additional method in case of change of props i.e. **componentWillReceiveProps**) of the React cycle and compare the changes, changes are done accordingly and those changes are reflected in the UI. 
  - React provides a declarative API so that you don’t have to worry about exactly what changes on every update. This process is called reconciliation. 
  - The reconciliation process is the way React updates the DOM, by making changes to the component based on the change in state.
  - When the request to setState() is triggered, React creates a new tree containing the reactive elements in the component (along with the updated state).

>   - React knows which changes to implement and will only update the parts of the DOM where necessary.

Reviewer 3: This is unfortunately too much detail at this stage of learning. If you want to link to a detailed explanation in the Supplemental Materials, feel free. Otherwise, all you have to mention is that calling `setState` will cause a re-render.

- Let us sum up the flow with an example.

>   - Say we have a component ```<Button />``` which takes its background color as a prop from parent component, which is associated in state of parent component. 

Reviewer 2: ```suggestion
  - Say we have a component ```<Button />``` which takes its background color as a `prop` from parent component, which is associated in state of parent component. 
```

  ```javascript
  state = {}
  ```
- Initialize state inside class constructor.
  ```javascript
  constructor(){
    this.state = {
    message: "Hello world"
    }
  }
  ```

- State should not be mutated directly i.e. a copy of state should be created and then updated.
  - Incorrect way -
    ```javascript
    this.state.message = "new message"
    ```
  - Correct way -
    ```javascript
        this.setState({
          message: "new message"
        })
    ```

> - The important thing to remember about setState is that is causes the component to re-render. Well, in introduction to React, we have learnt that React follows a special path called the React cycle. So, when state is updated using setState, React goes into the **React cycle** and then the changes are compared from the previous **React virtual DOM element**, changes are done accordingly and those changes are reflected in the UI.

Reviewer 2: ```suggestion
- The important thing to remember about setState is that is causes the component to re-render. Well, in introduction to React, we have learnt that React follows a special path called the React cycle. So, when state is updated using `setState`, React goes into the **React cycle** and then the changes are compared from the previous **React virtual DOM element**, changes are done accordingly and those changes are reflected in the UI.
```

> **Now let us create an increment/decrement counter which includes click functionality and displaying that value. When we click on the increment button, 1 is added to the previous value and decrement button will subtract the number by 1. Initially the number will be 0.**

Reviewer 2: Making the entire line bold doesn't look good.

5. So, let us create a function which will run when increment button is clicked. To increment the value in the state, we will take the previous value and add 1 to it by using *setState* function.
```javascript
addValue = () => {
  this.setState({
	  value: this.state.value + 1
  })
}
```
**Note:- setState is a predefined function in React which is used to update the state. The setState function has two function defination i.e. it can either take object as a parameter or a function as a parameter.** 
- In the above instance of setState, an object has been passed as a parameter.
- In case where previous state of React is used to update the new state(just like above instance), the correct way is to pass a function as a parameter which itself has previous state parameter and that can be used to update state.

```javascript
addValue = () => {
  this.setState((prevState) => {
    return {
      value: prevState.value +1
    }
  })
}
```
- For now let us use the object instance only.

6. Similarly we will create a function which will decrement the value by 1.

```javascript
subtractValue = () => {
  this.setState({

> 	  value: this.state.value + 1

Reviewer 2: Should this be `value: this.state.value - 1` ?
Correct me if I'm wrong.

Author: No, you are right. Typo mistake.

  })
}
```
7. The functions have been created, we just need to set the attributes of the button we have created i.e. associate the functions we have created to the buttons.
```html
<button onClick={this.addValue}>Increment</button>
<button onClick={this.subtractValue}>Decrement</button>
```
That's all. We have completed our first application in React using React state. To see the complete code -

> [Codepen](https://codepen.io/ashish24_nagpal/pen/jObzXzM)

Reviewer 2: Why are these in two different line ?

<p>{this.state.value}</p>
```

> 4. That is all which will be displayed on the page. Now we need to implement the functionality when the button are clicked.

Reviewer 2: ```suggestion
4. That is all which will be displayed on the page. Now we need to implement the functionality when the buttons are clicked.
```

5. So, let us create a function which will run when increment button is clicked. To increment the value in the state, we will take the previous value and add 1 to it by using _setState_ function.

```javascript
addValue = () => {
  this.setState({
    value: this.state.value + 1
  });
};
```

**Note:- setState is a predefined function in React which is used to update the state. The setState function has two function defination i.e. it can either take object as a parameter or a function as a parameter.**

- In the above instance of setState, an object has been passed as a parameter.
- In case where previous state of React is used to update the new state(just like above instance), the correct way is to pass a function as a parameter which itself has previous state parameter and that can be used to update state.

```javascript
addValue = () => {
  this.setState((prevState) => {
    return {
      value: prevState.value + 1
    };
  });
};
```

For now let us use the object instance only.

6. Similarly we will create a function which will decrement the value by 1.

```javascript
subtractValue = () => {
  this.setState({

>     value: this.state.value + 1

Reviewer 2: ```suggestion
    value: this.state.value - 1
```

  })
}
```

> 7. The functions have been created, we just need to associate the functions we have created to the buttons i.e. set the *onClick* attribute of the button which we have created.

Reviewer 2: ```suggestion
7. The functions have been created, we just need to associate the functions we have created to the buttons i.e. set the *onClick* attribute to the button which we have created.
```

```html
<button onClick={this.addValue}>Increment</button>
<button onClick={this.subtractValue}>Decrement</button>
```
8. Let us extend this example by adding lower and upper limit to the value. If the value is *0* then  we cannot reduce the value and when the value becomes *10*, we should not be able to increment the value.  
This can be done by hiding the buttons on certain boundary value.  
- If the value is 0 then decrement button should not be displayed.

> - Similarly we have to hid increment button when value becomes 10.  

Reviewer 2: ```suggestion
- Similarly we have to hide increment button when value becomes 10.  
```

      </div>
    );
  }
}
```
- App Component
```javascript
class App extends React.Component {
  constructor() { //Constructor
    super();
    this.state = { 
      message: "message" // message property defined in state
    };
  }
	// This function will change the state of App Component
  changeState = () => {
    this.setState({
      message: "New Message"
    })
  }
  render() {
    return (
      <div>
       <button type="button" onClick={this.changeState} > Change App Comp State</button>
      <Message message={this.state.message} />
			{/* Message Component - prop passed as message */}
      </div>
    );
  }
}

```
Now let us click this button in sequence defined - 
1. Click on `Change App Comp State` button. We see that the value of `message` has been changed to `new message` as the state of `App` component changes, the component is re-rendered and the value is passed as prop to `Message` component and the change is reflected.
2. Now click on `Change Message Comp state` button. We see that the `value` is now `world`. Again the component changes its state and so it is re-rendered.

> 3. Now click on `Change prop by Message Comp` button. We see that no change happens on the UI.

Reviewer 2: ```suggestion
3. Now click on `Change prop by Message Comp` button. We see that no change happens to the UI.
```

4. Now again click on `Change Message Comp state` button. We see that the value of `message` prop has been changed to `Change Prop message`.

How did this happen ?  
Well, when we clicked on `Change prop by Message Comp` button the value of prop changed. But since by changing its own prop component does not re-render itself, hence the changes were not displayed. Although the value of prop was changed to *Change Prop message*, it was not reflected. When we clicked on `Change Message Comp state` button, the state updated and so the component is re-render, therefore the prop changes are also reflected on the UI. 

One more thing to see here is that the prop `message` sent by `App` component had the value *message* first and then changed to *new message*. But now `message` prop has value *Change Prop message*. This shows inconsistency, therefore components follow the rule of pure functions which state that the props (parameters) received by component should not be updated.

> To see the complete code [Codepen](https://codepen.io/ashish24_nagpal/pen/oNjyeeo?editors=1111)

Reviewer 2: ```suggestion
To see the complete code [Codepen](https://codepen.io/ashish24_nagpal/pen/oNjyeeo)
```

     ```	

### Independent Practice

> - Create an increment/decrement counter which includes click functionality. On click of increment button adds 1 to the value and decrement will subtract the number by 1. Initially the number will be 0.

Reviewer 2: ```suggestion
- Create an increment/decrement counter which includes click functionality. On click of increment button adds 1 to the value and decrement button will subtract the number by 1. Initially the number will be 0.
```
